precompiles/xcm/src/lib.rs

```diff
@@ -365,16 +433,567 @@ where
+
+        if fee_asset_item as usize > assets.len() {
+            return Err(revert("Bad fee index."));
+        }
+
+        let assets: MultiAssets = assets
+            .iter()
+            .cloned()
+            .zip(amounts.iter().cloned())
+            .map(Into::into)
+            .collect::<Vec<MultiAsset>>()
+            .into();
+
+        // Build call with origin.
+        let origin = Some(Runtime::AddressMapping::into_account_id(
+            handle.context().caller,
+        ))
+        .into();
+
+        log::trace!(target: "xcm-precompile:assets_reserve_transfer", "Processed arguments: assets {:?}, amounts: {:?}, beneficiary: {:?}, destination: {:?}, fee_index: {}", assets, amounts, beneficiary, dest, fee_asset_item);
+
+        let call = pallet_xcm::Call::<Runtime>::reserve_transfer_assets {
+            dest: Box::new(dest.into()),
+            beneficiary: Box::new(beneficiary.into()),
+            assets: Box::new(assets.into()),
+            fee_asset_item,
+        };
+
+        // Dispatch a call.
+        RuntimeHelper::<Runtime>::try_dispatch(handle, origin, call)?;
+
+        Ok(succeed(EvmDataWriter::new().write(true).build()))
+    }
+
+    fn send_xcm(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+        let mut input = handle.read_input()?;
+        input.expect_arguments(3)?;
+
+        // Raw call arguments
+        let dest: MultiLocation = input.read::<MultiLocation>()?;
+        let xcm_call: Vec<u8> = input.read::<BoundedBytes<GetXcmSizeLimit>>()?.into();
+
+        log::trace!(target:"xcm-precompile::send_xcm", "Raw arguments: dest: {:?}, xcm_call: {:?}", dest, xcm_call);
+
+        let xcm = xcm::VersionedXcm::<()>::decode_all_with_depth_limit(
+            xcm::MAX_XCM_DECODE_DEPTH,
+            &mut xcm_call.as_slice(),
+        )
+        .map_err(|_| revert("Failed to decode xcm instructions"))?;
+
+        // Build call with origin.
+        let origin = Some(Runtime::AddressMapping::into_account_id(
+            handle.context().caller,
+        ))
+        .into();
+        let call = pallet_xcm::Call::<Runtime>::send {
+            dest: Box::new(dest.into()),
+            message: Box::new(xcm),
+        };
+        log::trace!(target: "xcm-send_xcm", "Processed arguments:  XCM call: {:?}", call);
+        // Dispatch a call.
+        RuntimeHelper::<Runtime>::try_dispatch(handle, origin, call)?;
+
+        Ok(succeed(EvmDataWriter::new().write(true).build()))
+    }
+
+    fn transfer(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+        let mut input = handle.read_input()?;
+        input.expect_arguments(5)?;
+
+        // Read call arguments
+        let currency_address = input.read::<Address>()?;
+        let amount_of_tokens = input
+            .read::<U256>()?
+            .try_into()
+            .map_err(|_| revert("error converting amount_of_tokens, maybe value too large"))?;
```

Comment:
It really confuses me. Why you used `try_into` for all conversions in xtokens related precompile impl but use manual conversion with other impl, and the error messages are not consistent? I assume it's copy & past without checking?

Reply:
Yes, I just used the legacy code that we were already using so didn't spend much refactoring it. I have made note of all your comments about the changes we need, I will update it in the follow up PR.

Reply:
TBH it's not professional to copy & past code here and there without double checking what code you used. Please avoid this kind of way of working in the future.

Reply:
I just used the code the same code for same function that we already have in production for a long time. It was not mindless copy paste. I wouldn't have been able to nitpick it like you.